WPWebChromeClient with File Chooser support for form upload input type

jd-alexander opened the pull request:
Fixes #14967

## Findings
The `WPWebChromeClient` does not currently override the `onShowFileChooser` method that allows for files to be uploaded when the file upload input type is utilized. 

Therefore the `WPWebChromeClient` was extended as `WPWebChromeClientWithFileChooser` within the main app, since the integration with `MediaUtils` from FluxC would be seamless. 

## Solution
 - The `WPWebChromeClient` implements the`onShowFileChooser` so that it now integrates with the `WPWebViewActivity ` to provide a file picking experience for any intents that utilize the `WebView` functionality. 
 
- I utilized the depreciated startForActivityResult because its usage was already present within the class. 

- I did not integrate with the MediaPicker Launcher because how the form chooser intent operates is different from how the API is structured therefore the effort to integrate with it did not seem worth it at the moment. The File Chooser supports other input modes such as Camera or Video Capture. Currently, this solution does not support this as it may be better to wait until [WordPress MediaPicker Android](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android)  is finalized so that a possible integration could be done. This solution will be shared there as a means of inquiring if there may be the possibility that the picker would prepare the API or clients for the use cases of the Webview's Form Chooser.  

- After completing the implementation, I thought about moving the logic into `WPWebChromeClient` inside the WordPress Utils class. I didn't get to that because I was trying to look at other groundskeeping issues. However, it seems like it would be worth the effort, I move it. However, there would need to be several changes to how the class is written to facilitate components that aren't accessible from the Utils library. 

## Testing

### Forms Plugin
Create a form utilzing corresponding MIME types of any of the file types in each category to verify that the File Picker is filtering MIME types appropriately. The functionality verification should also cover single vs multiple selection.

You can utilize https://file-examples.com/ to grab sample files that match the associated MIME types.

### Create Form From Scratch
1.  Install a custom plugin like CF7: [https://wordpress.org/plugins/contact-form-7/](https://wordpress.org/plugins/contact-form-7/)
2.  Edit the default form and make sure you add a [file upload field](https://contactform7.com/file-uploading-and-attachment/)
3.  Add the CF7 block to a post or page in Calypso.
4.  Tap the 3-dots, then "Preview" in the app. 
5.  Press Choose File on the file upload field.
6.  **Expect** that you are able to select the file using the System Picker.
7. **Optional** : Send the completed form to the server.

https://user-images.githubusercontent.com/1509205/141753102-ffb0375f-f98c-4a8e-b112-0458d6560a50.mov

### Import Prexisting Form 

1.  Install Caldera Forms: [https://wordpress.org/plugins/caldera-forms/](https://wordpress.org/plugins/caldera-forms/)
2. Import the [attached form](https://gist.github.com/jd-alexander/17e6d9ca323f133ad78b7dc10314e695).
3.  Add the Caldera Form block to a post or page in Calypso.
4.  Tap the 3-dots, then "Preview" in the app. 
5.  Press Choose File on the file upload fields within the form.
6.  **Expect** that you are able to select file (s) using the System Picker.
7. **Optional** : Send the completed form to the server.

N.B some MIME types aren't filtered as expected because the accepted types implementation in Caldera Forms does not function as expected.

https://user-images.githubusercontent.com/1509205/142083390-34b18505-3d79-497d-af21-92f3f208562d.mov

## Regression Notes

1. Potential unintended areas of impact

I don't see any potential unintended areas of impact with the changes introduced. The functionality of 
`WPWebChromeClient` was extended in a manner that doesn't interfere with any of it's existing behavior.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

To test I utilized two form building plugins that would be capable of creating the `input type` fields within a form that facilitate file uploads. To satisfy MIME type requirements, I carried out tests that combined different MIME types and selection params - single or multiselect. 

3. What automated tests I added (or what prevented me from doing so)

I didn't add any automated tests for this flow since we aren't currently currently testing `WPWebChromeClient`.

## Reviewing 
Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jd-alexander committed:
Added onActivityResult code that will be receiving URI results.

jd-alexander committed:
Extended the WPWebChromeClient to support file choosing.

jd-alexander committed:
Created a request code for the Web Chrome Client's File Picker result

jd-alexander committed:
Removed MediaPickerLauncher integration and utilized a simple approach.

jd-alexander committed:
removed unnecessary MediaPickerLauncher object.

jd-alexander committed:
Enhance the logic with the requestCode and did some cleanup.

jd-alexander committed:
Added support for resolving the mime type from an extension.

jd-alexander committed:
Improved the comment about getAcceptTypes and removed unused import.

jd-alexander committed:
The mFilePathCallback will always need a value to end the request.

jd-alexander committed:
Removed unused import.

jd-alexander committed:
Removed logic that limited EXTRA_MIME_TYPES usage for multiple inputs.

The EXTRA_MIME_TYPES property was only being set when the accepted types were more than one. This is a bug that has been corrected. 

The code was also formatted and comments improved.

jd-alexander committed:
Made the access modifier protected for compat with Jetpack Web View Act.

jd-alexander committed:
Ensure mFilePathCallback is not null

This is necessary because it is only set when onShowFileChooser() is called and the onActivityResult could be triggered by other functionality.

jd-alexander committed:
Added mFilePathCallback behavior explanation.

jd-alexander committed:
Added RELEASE-NOTES entry.

jd-alexander commented:
I just noticed that when I submit forms I am not getting the file upload even though it is attached. I don't think this should be an issue with the implementation but I will need to test again to make sure it has nothing to do with the code. If anyone wants to do a first pass, I have no problem with that 🙇🏾

jd-alexander commented:
So the implementation here is fine. I was able to successfully submit an attachment that reached my email. The issue is with the Contact Form 7 and Caldera Form plugins. They have extra configurations for file attachments. 

eg, in Contact Form 7, you need to supply the field name in the Mail settings. 

![plugin](https://user-images.githubusercontent.com/1509205/142130638-d3c1518f-8264-4050-afeb-0bf78eadb035.png)

For more details on this, you can visit this [tutorial](https://www.freakyjolly.com/resolved-contactform7-file-attachment-not-recieved-in-email/).

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/WPWebViewActivity.java:
```diff
@@ -957,4 +960,16 @@ public void onCookieProgressDialogCancelled() {
+        }
+    }
+
+    @Override public void startActivityForFileChooserResult(Intent intent, int requestCode) {
```
This `@Override` also should probably go one line up :)

jd-alexander replied:
Resolved here b0b011e

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/WPWebChromeClientWithFileChooser.java:
```diff
@@ -0,0 +1,119 @@
+        this.mOnShowFileChooserListener = onShowFileChooserListener;
+    }
+
+    @Override public boolean onShowFileChooser(WebView webView, ValueCallback<Uri[]> filePathCallback,
```
Looks like `@Override` crept down here :)

jd-alexander replied:
Resolved here b0b011e

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/WPWebChromeClientWithFileChooser.java:
```diff
@@ -0,0 +1,119 @@
+        // Sets an explicit MIME data type that states that all MIME types are supported.
+        intent.setType("*/*");
+        // Creates values for Intent.EXTRA_MIME_TYPES with the MIME types that are currently acceptable.
+        String[] resolvedMimeTypes = new String[acceptableMimeTypes.length];
```
I think we can use `acceptableMimeTypesLength` instead of `acceptableMimeTypes.length` here.

jd-alexander replied:
Makes sense 43a5c51

jd-alexander committed:
Merge branch 'develop' into issue-14967/web-preview-media-picker-integration

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/WPWebChromeClientWithFileChooser.java:
```diff
@@ -0,0 +1,119 @@
+             *  MIME types. The logic below is to accommodate the use cases by utilizing the extension to resolve
+             *  the appropriate Mime type with MediaUtils.getMimeTypeForExtension().
+             */
+            if (acceptableMimeType.contains(".")) {
```
If I understand this correctly, we check for dot here to detect mime type that is just an extension, eg. `.png`. Looks like regular mime-types can also have dots in them (eg. `application/vnd.ms-excel`), so maybe it makes sense to check if the `/` is absent (eg. `acceptableMimeType.contains(".") && !acceptableMimeType.contains("/")`)

jd-alexander replied:
Good find! 152424e

khaykov reviewed (commented):
Thanks for the investigation and a fix, @jd-alexander ! I confirmed that the file picker works, and the file is correctly uploaded using CF7. All the issues I had were related to the plugin, and not the implementation itself.

I also left a couple of minor comments in the code :)

jd-alexander committed:
Fixed the location of the @Override annotation on two functions.

jd-alexander committed:
Added logic that supports mime types such as application/vnd.ms-excel

This ensure that a fully qualified mime type is only created if we are sure that the input we got is a file extension and not a mime type that contains both a dot and a slash.

jd-alexander committed:
Utilized acceptableMimeTypesLength instead of the length property.

jd-alexander commented:
Thanks for the feedback @khaykov I am going to implement the changes and then do a final run thorough and then re-request a review from you 🙇🏾

jd-alexander committed:
Merge branch 'develop' into issue-14967/web-preview-media-picker-integration

jd-alexander committed:
Merge branch 'develop' into issue-14967/web-preview-media-picker-integration

khaykov reviewed (approved):
Thanks for the changes, @jd-alexander ! Looks good to me 👍
